### PR TITLE
Disparity processing

### PR DESCRIPTION
**Pull request by PascalleW**
to3D function implemented.
- Creates a vector of the 3D coordinates made from the disparity image.
- Added tests

### CODE REVIEW

File: catkin_ws/src/cucumber_msgs/include/cucumber_msgs/CucumberContainer.h [hunk 1]
```diff
@@ -71,7 +71,7 @@ class CucumberContainer {
 		 msg.width = width;
 		 msg.height = height;
 		 msg.curvature = curvature;
-		 std::vector<int> img_pos(image_x, image_y);
+		 std::vector<int> img_pos {image_x, image_y};
```
Comment (ChielBruin) [inline catkin_ws/src/cucumber_msgs/include/cucumber_msgs/CucumberContainer.h:74]: Good that you found this!

File: catkin_ws/src/kpr_image_processing/CMakeLists.txt [hunk 2]
```diff
@@ -67,9 +70,9 @@ add_dependencies(IdCB
 roslaunch_add_file_check(launch)
 
 ## Add gtest based cpp test target and link libraries
-# catkin_add_gtest(${PROJECT_NAME}-test test/test_image_processing.cpp)
+catkin_add_gtest(IdCB-test test/TEST_IdCB.cpp)
 # if(TARGET ${PROJECT_NAME}-test)
-#   target_link_libraries(${PROJECT_NAME}-test ${PROJECT_NAME})
+	target_link_libraries(IdCB-test ${catkin_LIBRARIES})
 # endif()
```
Comment (ChielBruin) [inline catkin_ws/src/kpr_image_processing/CMakeLists.txt]: The if and endif line can be removed

File: catkin_ws/src/kpr_image_processing/src/IdCB/IdCB.cpp [hunk 3]
```diff
@@ -13,14 +17,46 @@ std::map<ros::Time, stereo_msgs::DisparityImage> disparity;
...
 /**
  * Calculate the 3D properties of the cucumber from the given camera position and disparity image.
+ * Z = distance along the camera Z axis (in metres)
+ * f = focal length (in pixels)
+ * B = baseline (in metres)
+ * d = disparity (in pixels)
+ * Z = fB/d
+ * X = xZ/f (x is pixel-x-position)
+ * Y = yZ/f (y is pixel-y-position)
+ * Return: 3D postion cucumber(m) + width (m), height (m) and curvature
```
Comment (ChielBruin) [inline catkin_ws/src/kpr_image_processing/src/IdCB/IdCB.cpp]: Use @return and add @param tags for your input parameters

File: catkin_ws/src/kpr_image_processing/src/IdCB/IdCB.cpp [hunk 4]
```diff
@@ -13,14 +17,46 @@ std::map<ros::Time, stereo_msgs::DisparityImage> disparity;
...
+	float B = disparity.T;
+	float pixel_size = 4.65e-6;
+	float f = disparity.f;
+	int im = disparity.image.step*(y+48)+x*disparity.image.step/disparity.image.width;
+	if (im > disparity.image.data.size()){
+		ROS_ERROR("Out of Range");
+		return CucumberContainer(0, 0, 0, 0, -1);
+	}
+	int d = disparity.image.data[im];
+	if(camera != CAM_LEFT) {
```
Comment (ChielBruin) [inline catkin_ws/src/kpr_image_processing/src/IdCB/IdCB.cpp]: You might want to move this check to the start of the method for speed reasons

File: catkin_ws/src/kpr_image_processing/src/IdCB/IdCB.cpp [hunk 5]
```diff
@@ -13,14 +17,46 @@ std::map<ros::Time, stereo_msgs::DisparityImage> disparity;
...
+	float height = in.height*pixel_size;
+	float curvature = in.curvature;
+	Eigen::Vector4f v(X_cam,Y_cam,Z_cam,1);
+ 	//TODO translation and rotation degree
+	Eigen::Vector3f translate(6e-2,2e-2,47.4e-3-3.8e-3+14e-2);
+	float rotDeg = 30;
+	float rotRad = rotDeg*M_PI/180.0;
+	Eigen::Transform<float,3,Eigen::Affine> transform = Eigen::Translation3f(translate) * Eigen::AngleAxisf(rotRad,Eigen::Vector3f(1,0,0));
+	v = transform*v;
+	return CucumberContainer(v[0],v[1],v[2],width, height, curvature);
```
Comment (ChielBruin) [inline catkin_ws/src/kpr_image_processing/src/IdCB/IdCB.cpp]: The returned cucumber container does not include the original image position that is needed by the interface

File: catkin_ws/src/kpr_image_processing/test/TEST_IdCB.cpp [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+ * Run all the test cases.
+ */
+int main (int argc, char** argv) {
+	testing::InitGoogleTest(&argc, argv);
+	return RUN_ALL_TESTS();
+}
+
+/**
+ * Tests for to3D() function.
+to3D(cucumber_msgs::Cucumber in, int camera, stereo_msgs::DisparityImage disparity)
```
Comment (ChielBruin) [inline catkin_ws/src/kpr_image_processing/test/TEST_IdCB.cpp]: Add a * at the beginning of these two lines

File: catkin_ws/src/kpr_image_processing/test/TEST_IdCB.cpp [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+CucumberContainer c = CucumberContainer(x,y,width(cucumber),height,curve);
+ */
+stereo_msgs::DisparityImage CreateDisparity(int f, float T, int step, int width, int vecLength, int dataPoint, float dataPointValue){
```
Comment (ChielBruin) [inline catkin_ws/src/kpr_image_processing/test/TEST_IdCB.cpp:21]: This method is missing its documentation. The comment above is a general one about the tests below

File: catkin_ws/src/kpr_image_processing/test/TEST_IdCB.cpp [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+	d.T = T;
+	sensor_msgs::Image i = sensor_msgs::Image();
+	i.step = step;
+	i.width = width;
+	std::vector<uint8_t> arr(vecLength);
+	arr[dataPoint] = dataPointValue;
+	i.data = arr;
+	d.image = i;
+	return d;
+}
```
Comment (ChielBruin) [inline catkin_ws/src/kpr_image_processing/test/TEST_IdCB.cpp:33]: Add newline

**Review by ChielBruin — CHANGES_REQUESTED** [review]
The code looks good, but I had some tiny things I like to see changed regarding formatting and documentation

**Review by ChielBruin — APPROVED** [review]

File: catkin_ws/src/kpr_image_processing/src/IdCB/IdCB.cpp [hunk 9]
```diff
@@ -13,14 +17,46 @@ std::map<ros::Time, stereo_msgs::DisparityImage> disparity;
...
-	} else {
-		return tmp;
+CucumberContainer to3D(cucumber_msgs::Cucumber in, int camera, stereo_msgs::DisparityImage disparity) {
+	CucumberContainer tmp = CucumberContainer(in);
+	if(camera != CAM_LEFT) {
+		ROS_WARN("Right camera not supported");
+		return CucumberContainer(0, 0, 0, 0, -1);
+	}
+	int x = in.image_stem_position[0];
+	int y = in.image_stem_position[1];
```
Comment (MaxGroenenboom) [inline catkin_ws/src/kpr_image_processing/src/IdCB/IdCB.cpp]: This tab does not belong here!

File: catkin_ws/src/kpr_image_processing/src/IdCB/IdCB.cpp [hunk 10]
```diff
@@ -13,14 +17,46 @@ std::map<ros::Time, stereo_msgs::DisparityImage> disparity;
...
+	Eigen::Vector4f v(X_cam,Y_cam,Z_cam,1);
+ 	//TODO translation and rotation degree
+	Eigen::Vector3f translate(6e-2,2e-2,47.4e-3-3.8e-3+14e-2);
+	float rotDeg = 30;
+	float rotRad = rotDeg*M_PI/180.0;
+	Eigen::Transform<float,3,Eigen::Affine> transform = Eigen::Translation3f(translate) * Eigen::AngleAxisf(rotRad,Eigen::Vector3f(1,0,0));
+	v = transform*v;
+	CucumberContainer res = CucumberContainer(v[0],v[1],v[2],width, height, curvature);
+	res.setImagePosition(x,y);
+	return res;
```
Comment (MaxGroenenboom) [inline catkin_ws/src/kpr_image_processing/src/IdCB/IdCB.cpp:63]: Maybe add some blank lines between lines that do similar things to improve readability

**Review by MaxGroenenboom — CHANGES_REQUESTED** [review]

**Review by MaxGroenenboom — APPROVED** [review]

**Pull request merged**